### ISSUE TITLE
Question about the scale_constant in the source code

### ISSUE THREAD

**Issue by Thedatababbler** [QingruZhang/PASTA#5]
Hi, thank you for your great job. I've read both the paper and the source code in pasta.py. I've noticed that the scaling coefficient alpha in the paper is multiplied by the Attention value and come with a renormalization operation. However, in the implementation in pasta.py line 167-172, I found the following code:
```
        for bi, (ti,tj) in enumerate(token_range.tolist()):
            if self.scale_position == "include":
                attention_mask[bi, head_idx, :, ti:tj] += scale_constant
            else:
                attention_mask[bi, head_idx, :, :ti] += scale_constant
                attention_mask[bi, head_idx, :, tj:input_len] += scale_constant
```
If my understanding is correct, you are only doing additive steering, rather than the scaling coefficient approach you described in your code, and I found that there is no normalization operation in the code neither. Would you mind to elaborate the reason you use this implementation instead of the one you mentioned in your paper? Thank you.

**Comment by Thedatababbler** [comment]
Oh, sorry I just realized you applied additive scaling on the attention_mask, so it will be scaling coefficients when the attention value multiplies with the attention_mask. But, I still wonder where did you do the normalization part? Thanks.

**Comment by QingruZhang** [comment]
Hi, sorry for the late reply. We refactored from the original implementation and applied the adding constant by attention_mask. Adding $\log\alpha$ is equivalent to multiplying $\alpha$. Since we add on the inner product before softmax, the renormalization is no longer needed as the following softmax will do that.

**Comment by Thedatababbler** [comment]
Ok, great. Thank you for your explanation and nice work.

**Comment by SydCS** [comment]
Hi!
I have a question: why would +scale_constant at the attention_mask of excluded tokens help reduce attention on these tokens? Why do not their attn_weights go up after this op?
Thanks if someone could answer.

updata: I see. These is a log().
